crates/wasmtime/src/runtime/code.rs

```diff
@@ -100,3 +284,180 @@ impl From<Arc<ComponentTypes>> for Types {
         Types::Component(types)
     }
 }
+
+/// A `Store`-local instance of code.
+///
+/// This type encapsulates executable code within the context of
+/// instantiation in a single store. It may be an unmodified pointer
+/// to the read-only original code, or it may be locally patched for
+/// this store due to debugging or instrumentation settings.
+///
+/// Most things that the runtime will want to do with a module's image
+/// will require the `EngineCode` -- all metadata lives there. The
+/// `StoreCode` is solely responsible for the executable machine code.
+///
+/// This type is designed to be uniquely owned, unlike `EngineCode`
+/// above. The runtime data structures will have scattered `Arc`s to
+/// the `EngineCode` but only one
+pub struct StoreCode(StoreCodeStorage);
+
+enum StoreCodeStorage {
+    Shared(Arc<CodeMemory>),
+    /// Private copy of the given code memory.
+    ///
+    /// This is the only reference to this CodeMemory. The StoreCode
+    /// is owned directly by the Store's ModuleRegistry.
+    Private(Box<CodeMemory>),
+}
+
+impl StoreCode {
+    /// Create a new StoreCode for a given Store from a given
+    /// EngineCode, given the engine's settings (tunables).
+    pub fn new(engine: &Engine, engine_code: &Arc<EngineCode>) -> Result<Self> {
+        // Enabled guest-debugging causes us to allocate private
+        // copies of code in every store, to allow individual enabling
+        // of breakpoints (by code patching) independently in each
+        // one.
+        let code = if engine.tunables().debug_guest {
+            let mut private_copy = engine_code.original_code.deep_clone(engine)?;
```

Comment:
As a follow-up and/or pending issue, it might be nice to only clone the text section here as opposed to anything else since everything else in the module is irrelevant for the cloning purpose. Modules can often be 2x larger than the text section for the `.wasmtime.addrmap` section alone so we might have some nice wins cutting down on the size. My hope is that this'd be pretty easy refactor to tack on in the future though where the only caveat would be to redirect some accesses of metadata to the `EngineCode` layer, if any, rather than using it through `StoreCode` (and this may already very well be solved entirely in the refactorings of this PR)